test/common/upstream/edf_scheduler_test.cc

```diff
@@ -134,6 +171,113 @@ TEST(EdfSchedulerTest, ManyPeekahead) {
   }
 }
 
-} // namespace
+// Validates that creating a scheduler using the createWithPicks (with 0 picks)
+// is equal to creating an empty scheduler and adding entries one after the other.
+TEST_F(EdfSchedulerTest, EqualityAfterCreateEmpty) {
+  constexpr uint32_t num_entries = 128;
+  std::vector<std::shared_ptr<uint32_t>> entries;
+  entries.reserve(num_entries);
+
+  // Populate sched1 one entry after the other.
+  EdfScheduler<uint32_t> sched1;
+  for (uint32_t i = 0; i < num_entries; ++i) {
+    entries.emplace_back(std::make_shared<uint32_t>(i + 1));
+    sched1.add(i + 1, entries.back());
+  }
+
+  EdfScheduler<uint32_t> sched2 = EdfScheduler<uint32_t>::createWithPicks(
+      entries, [](const double& w) { return w; }, 0);
+
+  compareEdfScehdulers(sched1, sched2);
+}
+
+// Validates that creating a scheduler using the createWithPicks (with 5 picks)
+// is equal to creating an empty scheduler and adding entries one after the other,
+// and then performing some number of picks.
+TEST_F(EdfSchedulerTest, EqualityAfterCreateWithPicks) {
+  constexpr uint32_t num_entries = 128;
+  // Use double-precision weights from the range [0.01, 100.5].
+  // Using different weights to avoid a case where entries with the same weight
+  // will be chosen in different order.
+  std::vector<std::shared_ptr<double>> entries;
+  entries.reserve(num_entries);
+  for (uint32_t i = 0; i < num_entries; ++i) {
+    const double entry_weight = (100.5 - 0.01) / num_entries * i + 0.01;
+    entries.emplace_back(std::make_shared<double>(entry_weight));
+  }
+
+  const std::vector<uint32_t> all_picks{5, 140, 501, 123456, 894571};
+  for (const auto picks : all_picks) {
+    // Populate sched1 one entry after the other.
+    EdfScheduler<double> sched1;
+    for (uint32_t i = 0; i < num_entries; ++i) {
+      sched1.add(*entries[i], entries[i]);
+    }
+    // Perform the picks on sched1.
+    for (uint32_t i = 0; i < picks; ++i) {
+      sched1.pickAndAdd([](const double& w) { return w; });
+    }
+
+    // Create sched2 with pre-built and pre-picked entries.
+    EdfScheduler<double> sched2 = EdfScheduler<double>::createWithPicks(
+        entries, [](const double& w) { return w; }, picks);
+
+    compareEdfScehdulers(sched1, sched2);
+  }
+}
+
+// Emulates first-pick scenarios for the given number of iterations by creates
+// a scheduler with the given weights and a random number of pre-picks, and
+// validates that the next pick of all the weights is close to the given weights.
+void firstPickTest(const std::vector<double> weights) {
+  TestRandomGenerator rand;
+  // To be able to converge to the expected weights, a decent number of iterations
+  // should be used. If the number of weights is large, the number of iterations
+  // should be larger than 10000.
+  constexpr uint64_t iterations = 100000;
```

Comment:
I think that having a single test that has many random values makes more sense in this case, because at the end of test, it compares the expected and observed values.